### PR TITLE
Template for 2d microscopy segmentation

### PR DESCRIPTION
**Pull request by hermancollin**
This script is still a WIP. It aims at reproducing the default SEM model used in ADS, which was orignally trained using ivadomed.

### CODE REVIEW

File: microscopy_segmentation/unet_train_monai.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    LoadImaged,
+    RandAffined,
+    Rand2DElasticd,
+    NormalizeIntensityd,
+    ResizeWithPadOrCropd
+)
+from monai.visualize import plot_2d_or_3d_image
+
+
+def main(working_dir):
```
Comment (jcohenadad) [inline microscopy_segmentation/unet_train_monai.py:38]: why making it a function (as opposed to a script)?

Comment (hermancollin) [reply microscopy_segmentation/unet_train_monai.py]: @jcohenadad agreed. In this context, there should not be any command-line parsing.

File: microscopy_segmentation/unet_train_monai.py [hunk 2]
```diff
@@ -158,7 +158,7 @@
 # original config: 150 epochs but stopped at epoch 117 due to early stopping
 num_epochs = 117
 # TODO: HMM... MAYBE REMOVE SIGMOID FROM THE LOSS? DOESN'T SEEM TO BE APPLIED IN IVADOMED
```
Comment (jcohenadad) [inline microscopy_segmentation/unet_train_monai.py:160]: If you're doing 2 classes (axon, myelin), then this would be considered a 3-class segmentation (with the background), and softmax would be the way to go (with one-hot encoder for the labels). Maybe [this script](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/main/train.py) would be useful for you. It also includes nice visu on [wandb](https://wandb.ai/jcohenadad/mouse-zurich/runs/gcd6g3e2)

Comment (hermancollin) [reply microscopy_segmentation/unet_train_monai.py:160]: @jcohenadad The 3-class formulation does make a lot of sense to me. Thank you for the script I'll have a look.

Softmax also seems to me to be the best option for multiclass. The reason I didn't go for this one is that the activation function used in our default ADS models was always sigmoid, and it worked fine. I see you take care of the one-hot encoding yourself [here](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/3d450751312aceb19ebc87278fb7aaf412262545/train.py#L213-L215). I need to do something similar.

+1 for wandb

Comment (jcohenadad) [reply microscopy_segmentation/unet_train_monai.py:160]: hum, maybe @mariehbourget can chip in to give us some insights as to why sigmoid vs softmax was chosen in the previous implementation.

Comment (mariehbourget) [reply microscopy_segmentation/unet_train_monai.py:160]: Hello!
The sigmoid function was chosen in the models I trained with ivadomed because it was consistently giving better results than softmax (both quantitative and qualitative) with otherwise the same hyperparameters.
However, with more time, I think I would have tried to optimize other hyperparameters to stick with softmax because it makes more sense in multiclass to have an activation function that returns true probabilities (adding up to 1).

As for the number of `out_channels`, ivadomed was adding the 3rd channel (background) automatically on 2+ class segmentation tasks and the final prediction values were extracted from the results of sigmoid (or softmax) on the 3 classes with `binarize_maxpooling` from the config file.
Long story short, without `binarize_maxpooling`, a single-class segmentation task did not need to add the background channel so `in_channel = out_channel = 1`, whereas 2+ class segmentation needed the added background channel to compute the prediction.

Comment (jcohenadad) [reply microscopy_segmentation/unet_train_monai.py:160]: thank you so much @mariehbourget !! This is very helpful
